nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_constructor.rs

```diff
@@ -161,26 +165,68 @@ impl StringConstructor {
     /// the rest parameter `codePoints`.
     fn from_code_point(
-        _agent: &mut Agent,
+        agent: &mut Agent,
         _this_value: Value,
-        _arguments: ArgumentsList,
-        _gc: GcScope,
+        code_points: ArgumentsList,
+        mut gc: GcScope,
     ) -> JsResult<Value> {
+        // 3. Assert: If codePoints is empty, then result is the empty String.
+        if code_points.is_empty() {
+            return Ok(String::EMPTY_STRING.into_value());
+        }
+        // fast path: only a single valid code unit
+        if code_points.len() == 1 && code_points.first().unwrap().is_integer() {
+            // a. Let nextCP be ? ToNumber(next).
+            let next_cp = to_number(agent, code_points[0].into_value(), gc.reborrow())?.unbind();
+            // b. If IsIntegralNumber(nextCP) is false, throw a RangeError exception.
+            // c. If ℝ(nextCP) < 0 or ℝ(nextCP) > 0x10FFFF, throw a RangeError exception.
+            let next_cp = next_cp.into_i64(agent);
+            if next_cp > 0x10FFFF {
+                return Err(agent.throw_exception(
+                    ExceptionType::RangeError,
+                    format!("{:?} is not a valid code point", next_cp),
+                    gc.nogc(),
+                ));
+            }
+            // d. Set result to the string-concatenation of result and UTF16EncodeCodePoint(ℝ(nextCP)).
+            if let Some(cu) = char::from_u32(next_cp as u32) {
+                // 4. Return result.
+                return Ok(SmallString::from(cu).into());
+            }
+        };
         // 1. Let result be the empty String.
+        let mut result = std::string::String::new();
         // 2. For each element next of codePoints, do
-        //     a. Let nextCP be ? ToNumber(next).
-        //     b. If IsIntegralNumber(nextCP) is false, throw a RangeError exception.
-        //     c. If ℝ(nextCP) < 0 or ℝ(nextCP) > 0x10FFFF, throw a RangeError exception.
-        //     d. Set result to the string-concatenation of result and UTF16EncodeCodePoint(ℝ(nextCP)).
-        // 3. Assert: If codePoints is empty, then result is the empty String.
+        for next in code_points.iter() {
+            // a. Let nextCP be ? ToNumber(next).
+            let next_cp = to_number(agent, next.into_value(), gc.reborrow())?.unbind();
+            // b. If IsIntegralNumber(nextCP) is false, throw a RangeError exception.
+            if !is_integral_number(agent, next_cp, gc.reborrow()) {
+                return Err(agent.throw_exception(
+                    ExceptionType::RangeError,
+                    format!("{:?} is not a valid code point", next_cp.to_real(agent)),
+                    gc.nogc(),
+                ));
+            }
+            // c. If ℝ(nextCP) < 0 or ℝ(nextCP) > 0x10FFFF, throw a RangeError exception.
+            let next_cp = to_uint32_number(agent, next_cp);
+            if next_cp > 0x10FFFF {
```

Comment:
(continuing the above comment) And then this should fix the issue you had with -1:

```suggestion
            if next_cp < 0 || next_cp > 0x10FFFF {
```